src/utils/strbuf/strbuf.c

```diff
@@ -289,3 +289,47 @@ int strbuf_print_escaped(strbuf_t *buf, char const *s, char const *need_escape,
 
   return 0;
 }
+
+int strbuf_print_restricted(strbuf_t *buf, char const *s, char const *accept,
+                            char replace_char) {
+  if (buf == NULL || s == NULL || accept == NULL || accept[0] == 0 ||
+      replace_char == 0) {
+    return EINVAL;
+  }
+  if (strchr(accept, replace_char) == NULL) {
+    return EINVAL;
+  }
+
+  size_t s_len = strlen(s);
+  if (s_len == 0) {
+    return strbuf_print(buf, s);
+  }
+
+  while (s_len > 0) {
+    size_t valid_len = strspn(s, accept);
+    if (valid_len == s_len) {
+      return strbuf_print(buf, s);
+    }
+    if (valid_len != 0) {
+      int status = strbuf_printn(buf, s, valid_len);
+      if (status != 0) {
+        return status;
+      }
+
+      s += valid_len;
+      s_len -= valid_len;
+      continue;
+    }
+
+    char tmp[] = {replace_char, 0};
+    int status = strbuf_print(buf, tmp);
+    if (status != 0) {
+      return status;
+    }
+
+    s++;
+    s_len--;
+  }
```

Comment:
The loop body was a bit hard to reason about, so I have rewritten this using a "copy and replace" approach. Thanks for the suggestion.

Accepting a callback function sounds like a [YAGNI smell](https://testing.googleblog.com/2017/08/code-health-eliminate-yagni-smells.html). We don't need them.

On optimizing this function:

One of my guiding principles is: **Never optimize without profiling.** My experience has been that the bottleneck is never where I expect it to be. In 9 out of 10 cases, the bottleneck looks totally innocuous at first glance. In the past I have seen "compare" functions (for lookup/sorting) to be the bottleneck, or once a `strncpy()` call, because it zeros the entire destination buffer, which caused lots of page faults due to an oversized buffer. This function, I'm quite confident, is not going to be a bottleneck anytime soon. It's simply not central enough.

Callback functions tend to be slow also: they cannot be inlined, the compiler must use external calling conventions, and they break branch prediction.

`isalnum` and friends depend on the locale – I wouldn't be surprised if they were not actually as cheap as one might think. Again, profiling is key.

Reply:
Definitely agree on profiling (so much that I've written CPU+DSP profiler for an emulator :-)). The main issue was indeed the code giving "too complicated" smell, callback suggestion came only after that...

Hardcoding the checks instead of using `isalnum()` could be faster (I guess due to latter using locale specific lookup) according to this: https://stackoverflow.com/questions/5727307/c-isalpha-lookup-table

I don't remember seeing `strspn()` inlined by compiler, unlike more common string functions.  And I think inlining happens mainly for fixed sized inputs, which was not the case here.